File: scripts/strand_guessing.py
#!/usr/bin/env python3

# Author: serine
# Date: 10/11/2016
# Last updated: 12/12/2018

import os
import sys
import math
import argparse
import pandas as pd

parser = argparse.ArgumentParser(
    usage='%(prog)s --logs_dir <path/to/log_file> --samples_sheet <path/to/samples_sheet>',
    description="This script summarises log files information into html table",
    add_help=True
)
parser.add_argument('-d',
                    '--logs_dir',
                    metavar='FILE',
                    required=True,
                    help="specify directory with your BAM files, which should also hold "
                         "`*Log.final.out` files"
                    )
parser.add_argument('-s',
                    '--samples_sheet',
                    metavar='FILE',
                    required=True,
                    help="path to samplesSheet.txt file, format "
                         "old_prefix\tnew_prefix"
                    )
parser.add_argument('-p',
                    '--pattern',
                    metavar='STR',
                    default=".summary",
                    help="specify ending of your log files, e.g featureCounts is .summary"
)
parser.add_argument('-o',
                    '--outdir',
                    metavar='PATH',
                    default=".",
                    help="specify directory for the output, two files will be created"
                         "'strandInfoAll.txt' and 'strandInfoGuess.txt'"
                         " default currect directory"
)

args = parser.parse_args()
logs_dir = args.logs_dir
samples_sheet = args.samples_sheet
pattern = args.pattern
outdir = args.outdir

ss = open(samples_sheet).read().split("\n")
ss = [s.split("\t")[1] for s in ss if s]
sorted(ss, key=len)
ss = ss[::-1]

# Two files out
#  - strandInfoGuess.txt
#  - strandInfoAll.txt

# s_name = "Status"
# assign = "Assigned"
# ambig = "Unassigned_Ambiguity"
# multi_map = "Unassigned_MultiMapping"
# no_feat = "Unassigned_NoFeatures"
# unmaped = "Unassigned_Unmapped"
# un_map_qual = "Unassigned_MappingQuality"
# un_frag_len = "Unassigned_FragmentLength"
# chimer = "Unassigned_Chimera"
# secondary = "Unassigned_Secondary"
# non_junc = "Unassigned_Nonjunction"
# dups = "Unassigned_Duplicate"
#
# keys = [s_name,
#        assign,
#        ambig,
#        multi_map,
#        no_feat,
#        unmaped,
#        un_map_qual,
#        un_frag_len,
#        chimer,
#        secondary,
#        non_junc,
#        dups
#        ]

list_of_files = os.listdir(logs_dir)
data_dict = {}

# def sign(x):
#    if x >= 0:
#        return 1
#    return -1

def which_strand(strnd_val):

    # confidence test
    # non_strnd_test = (20-1)/float(20+1) #0.904
    # strnd_test = (55-45)/float(55+45)   #0.1
    strnd_test = float(19 - 1) / float(20 + 1)  # 0.857
    non_strnd_test = float(55 - 45) / float(55 + 45)  # 0.1

    if abs(strnd_val) > strnd_test:
        # print("Data is stranded %s" % sign(strnd_val))
        if strnd_val >= 0:
            #return "ForwardStrandedCounts,%s,0" % str(strnd_val)
            return ["ForwardStrandedCounts", strnd_val, 0]

        #return "ReverseStrandedCounts,%s,0" % str(strnd_val)
        return ["ReverseStrandedCounts", strnd_val, 0]

    elif abs(strnd_val) < non_strnd_test:
        # print("Data is non stranded %s" % sign(strnd_val))
        #return "NonStrandedCounts,%s,0" % str(strnd_val)
        return ["NonStrandedCounts", strnd_val, 0]
    # NOTE default to non stranded counts, should be ok to get through RNAsik run
    elif non_strnd_test < abs(strnd_val) < strnd_test:
        # print("It is hard to guess what strand the data is %s" % sign(strnd_val))
        return ["NonStrandedCounts", strnd_val, 1]
    else:
        sys.exit("ERROR: This should not happend")
        # return "NonStrandedCounts,1"

def get_name(raw_name, ss):
    #TODO this can be buggy when column name contains forward slash (/)
    # this function returns None when this is the case. Forward slash should be allowed in the column name
    base_name = os.path.basename(raw_name)
    for name in ss:
        if base_name.startswith(name):
            return name

    #TODO need to raise an error here if
    # it didn't exit in the for loop.
    #raise NoNameError: "Can't happen"

def get_df(f, name):
    df = pd.read_csv(f, sep = "\t")
    df2 = df.rename(columns=lambda x: get_name(x, ss), inplace=False)
    tmp = df2.iloc[0].to_frame().reset_index()
    df3 = tmp.rename(columns={tmp.columns[0]: "sample", tmp.columns[1]: name})
    # dropping second line, not too sure why, fucking pandas
    df3.drop(df3.index[:1], inplace=True)

    return df3

dfs = []

for text_file in list_of_files:
    if text_file.endswith(pattern):
        name = text_file.split('.')[0]
        full_path = os.path.join(logs_dir, text_file)

        dfs.append(get_df(full_path, name))

tmp = pd.merge(dfs[0], dfs[1], how = "left", on = 'sample')
df = pd.merge(tmp, dfs[2], how = "left", on = 'sample')

df['frac'] = (df['ForwardStrandedCounts'] - df['ReverseStrandedCounts']) / (df['ForwardStrandedCounts'] + df['ReverseStrandedCounts'])

#NOTE can return sorted table, but not sure if that will
# generically useful? one can use unix sort if needed
#df.sort_values(by=['frac'], inplace = True)

if not os.path.isdir(outdir) and not os.path.exists(outdir):
    sys.exit("ERROR: %s doesn't exists, make sure it does" % outdir)

out_fn_all = os.path.join(outdir, "strandInfoAll.txt")
out_fn_guess = os.path.join(outdir, "strandInfoGuess.txt")

df.to_csv(out_fn_all, sep = "\t", index = False)

df2 = df.drop(['sample', 'frac'], axis = 1)
df3 = df2.sum(axis = 0)
tot_frac = (df3['ForwardStrandedCounts'] - df3['ReverseStrandedCounts']) / (df3['ForwardStrandedCounts'] + df3['ReverseStrandedCounts'])

df_sum = pd.DataFrame([which_strand(tot_frac)])
df_sum.to_csv(out_fn_guess, sep = ",", index = False, header = False)

File: scripts/strand_guessing.R
#!/usr/bin/env Rscript

# NOTES:
#
# Column table 'Status Sample Value' where Status value is one of:
#
# - Assigned
# - Unassigned_Ambiguity
# - Unassigned_Chimera
# - Unassigned_Duplicate
# - Unassigned_FragmentLength
# - Unassigned_MappingQuality
# - Unassigned_MultiMapping
# - Unassigned_NoFeatures
# - Unassigned_NonSplit
# - Unassigned_Overlapping_Length
# - Unassigned_Secondary
# - Unassigned_Unmapped
#
# > res5: res2 data.frame with added column strand_sore, as per cal_strand_score function
# This data.frame is full of information about reads distribution and mappibility to features

library(readr)
library(dplyr)
library(tidyr)
library(purrr)

args = commandArgs(trailingOnly=TRUE)

#TODO can instead set it to default to countFiles/
if (length(args) == 0 ) {
  stop("Specify counts directory", call.=FALSE)
}

counts_dir = args[1]

if(!dir.exists(counts_dir)) {
  msg <- paste0("Directory doesn't exist -> ", counts_dir, "\n")
  stop(msg)
}

fns1 <- list.files(counts_dir, pattern = "txt.summary$", full.names = TRUE)

non_stranded_cnts <- fns1[grepl("_nonCnts.txt.summary", fns1)]
fwd_stranded_cnts <- fns1[grepl("_fwdCnts.txt.summary", fns1)]
rev_stranded_cnts <- fns1[grepl("_revCnts.txt.summary", fns1)]

fns2 <- list("NonStandedStats" = non_stranded_cnts,
             "ForwardStrandedStats" = fwd_stranded_cnts,
             "ReverseStrandedStats" = rev_stranded_cnts)

res1 <- fns2 %>% map(function(stats) map(stats, function(n) read_tsv(n)))

res2 <- res1 %>% map(function(dat) Reduce(function(x, y) left_join(x, y, by = "Status"), dat))

res3 <- res2 %>% map(function(dat) dat %>%
                                    gather(sample, stats, -Status) %>%
                                    mutate(sample = gsub("_sorted.(repaired|bam)$", "", basename(sample))))

names(res3) %>% map(function(n) {
                      fn_out <- paste0(counts_dir, "/", gsub("Stats", "Counts", n), ".tsv.summary")
		      dat <- res3[[n]]
		      ss <- unique(dat[["sample"]])
		      #TODO should assert that ss is a length of one
		      dat <- dat %>% select(-sample)
		      colnames(dat) <- c("Status", ss)
                      write_tsv(dat, fn_out)
                    })

res4 <- names(res3) %>% map(function(n) res3[[n]] %>% dplyr::rename(!!n := stats))

res5 <- Reduce(function(x, y) left_join(x, y, by = c("Status", "sample")), res4)

cal_strand_score <- function(fow, rev) {
  (fow - rev) / (fow + rev)
}

res6 <- res5 %>%
	  mutate(strand_score = cal_strand_score(ForwardStrandedStats, ReverseStrandedStats)) %>%
          arrange(Status, sample)
          #filter(abs(strand) > 0, !is.na(strand))

out1_fn <- paste0(counts_dir, "/strandInfoAll.tsv")
out2_fn <- paste0(counts_dir, "/strandInfoPerSample.tsv")
out3_fn <- paste0(counts_dir, "/strandInfoGuess.tsv")

res6 %>% write_tsv(out1_fn)

res6 %>%
  filter(Status == "Assigned") %>%
  write_tsv(out2_fn)

mean_strand_score <- res6 %>%
                       filter(Status == "Assigned") %>%
                       summarize(mean_strand_score = mean(strand_score)) %>%
                       select(mean_strand_score) %>%
	               unlist() %>%
	               unname()

strnd_test = (19 - 1) / (20 + 1)  # 0.857
non_strnd_test = (55 - 45) / (55 + 45)  # 0.1

out3 <- ""

if(abs(mean_strand_score) > strnd_test) {
  if(mean_strand_score >= 0) {
    out3 <- paste0("ForwardStrandedCounts,", mean_strand_score, ",0")
  } else {
    out3 <- paste0("ReverseStrandedCounts,", mean_strand_score, ",0")
  }
} else if(abs(mean_strand_score) < non_strnd_test) {
    out3 <- paste0("NonStrandedCounts,", mean_strand_score, ",0")
} else if(non_strnd_test < abs(mean_strand_score) | strnd_test > abs(mean_strand_score)) {
    out3 <- paste0("NonStrandedCounts,", mean_strand_score, ",1")
} else {
    stop("This shouldn't happen")
}

out3_header <- "strand_type,mean_score,exit_code"
gsub(",", "\t", out3_header) %>% write(out3_fn)
gsub(",", "\t", out3) %>% write(out3_fn, append = TRUE)

archive_dir <- paste0(counts_dir, "/.archive/")

if(!dir.exists(archive_dir)) {
  msg <- paste0("Archiving individual summary files -> ", archive_dir, "\n")
  cat(msg)
  dir.create(archive_dir, recursive = TRUE)
}

fns1 %>% map(function(n) file.rename(n, paste0(archive_dir, basename(n))))
